lib/mswp/mswp.rb

```diff
@@ -0,0 +1,148 @@
+require "md_array"
+
+module MSwp
+  class MSwp
+    class IllegalParameterSpecified < ArgumentError; end
+
+    class IllegalSizeSpecified < IllegalParameterSpecified; end
+
+    class IllegalMinesCountSpecified < IllegalParameterSpecified; end
+
+    class TooManyMinesCountSpecified < IllegalParameterSpecified; end
+
+    class Error < StandardError; end
+
+    class GameEnd < Error; end
+
+    class GameOver < GameEnd; end
+
+    class GameClear < GameEnd; end
+
+    def initialize(size, mines_count)
+      raise IllegalSizeSpecified unless size.all? { |v| v >= 1 }
+      raise IllegalMinesCountSpecified unless mines_count >= 0
+
+      all_cells_count = size.reduce(&:*)
+
+      raise TooManyMinesCountSpecified unless mines_count < all_cells_count
+
+      @cells = MdArray::MdArray.new(size) { Cell.new }
+      @size = size
+      @all_cells_count = all_cells_count
+      @mines_count = mines_count
+      @active = false
```

Comment:
Initialization of `@untouched_cells_count` and `@flagged_cells_count` should be additionally added here because they are publish as readers at `attr_reader`.